File: jan-20-25/sub.py
from typing import List

class Solution:
    def firstCompleteIndex(self, arr: List[int], mat: List[List[int]]) -> int:
        
        m, n = len(mat), len(mat[0])

        mp = [0] *(m * n + 1)

        for i in range(m):
            for j in range(n):

                mp[mat[i][j]] = i * n + j

        rows = [0] * m
        cols = [0] * n

        for i, val in enumerate(arr):
            hs = mp[val]
            rows[hs//n] += 1
            if rows[hs//n] == n:
                return i
            cols[hs % n] += 1
            if cols[hs % n] == m:
                return i

        return m * n - 1

File: jan-20-25/sub.java
class Solution {
    public int firstCompleteIndex(int[] arr, int[][] mat) {
        int m = mat.length, n = mat[0].length;
        int mp[] = new int[m * n + 1];

        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                mp[mat[i][j]] = i * n + j;
            }
        }

        int rows[] = new int[m], cols[] = new int[n];

        int len = arr.length, hs;
        for(int i = 0; i<len; i++){
            hs = mp[arr[i]];
            rows[hs/n]++;
            if(rows[hs/n] == n) return i;
            cols[hs % n]++;
            if(cols[hs % n] == m) return i;
        }

        return m * n - 1;
    }
}